Fixed Build Warnings in AOAB

aoab-contributor opened the pull request:
The goal of these changes is to build the project cleanly without warnings to make it easier to notice when subsequent changes introduce bugs that are statically detectable by the compiler.

This set of changes is meant to be largely a behavior-preserving refactor. As such, there was no attempt made to address potential architectural or code-flow issues.

For example, it might be possible to restructure the app to avoid potential NULL dereferences. But this fix does not attempt to do so - instead it makes explicit the existing behavior of throwing an exception by using the NULL-forgiving operator (!). Future updates may wish to address this short-coming by avoiding the possibility of NULL objects entirely.

aoab-contributor committed:
Updated variable and type names to avoid naming-rule violation warnings

Signed-off-by: aoab-contributor <124131985+aoab-contributor@users.noreply.github.com>

aoab-contributor committed:
Fixed warning: "Converting null literal or possible null value to non-nullable type."

Pure refactoring, no behavior affecting changes were made

Signed-off-by: aoab-contributor <124131985+aoab-contributor@users.noreply.github.com>

aoab-contributor committed:
Fixed warning: "Dereference of a possibly null reference."

Pure refactoring, no behavior affecting changes were made.

Signed-off-by: aoab-contributor <124131985+aoab-contributor@users.noreply.github.com>

aoab-contributor committed:
Fixed warning: "Possible null reference argument for parameter 'X' in 'Y'."

As much as possible, tried to avoid behavior-affecting modifying changes.

Signed-off-by: aoab-contributor <124131985+aoab-contributor@users.noreply.github.com>

aoab-contributor committed:
Fixed small Unix/Windows line-ending inconsistency

Signed-off-by: aoab-contributor <124131985+aoab-contributor@users.noreply.github.com>

aoab-contributor committed:
Fixed warning: The variable 'ex' is declared but never used, for cases of exception objects

For unused exception objects, ensured that they were used in subsequently thrown exceptions or in Console.WriteLn()

Signed-off-by: aoab-contributor <124131985+aoab-contributor@users.noreply.github.com>

aoab-contributor committed:
Fixed warning: The variable 'list' is assigned but its value is never used

Removed unused variable.

Signed-off-by: aoab-contributor <124131985+aoab-contributor@users.noreply.github.com>

aoab-contributor committed:
Fixed warning: "The local function X is declared but never used" during Release builds

Wrapped the following functions in conditional compilation flags:
 * AOABO/program.cs: async Task RedoJSON()
 * AOABO/program.cs: async Task SaveAll()
 * AOABO/program.cs: async Task Save(string filename, List<Volume> vols)
 * AOABO/program.cs: async Task AddChapter()
 * AOABO/program.cs: bool GetYN()
 * AOABO/program.cs: async Task CreateTables()
 * AOABO/program.cs: async Task PartChart(Chapter[] chapters, string name, bool partOne = false, bool partTwo = false, bool partThree = false, bool partFour = false, bool partFive = false)

Signed-off-by: aoab-contributor <124131985+aoab-contributor@users.noreply.github.com>

aoab-contributor committed:
Now that all warnings have been addressed, instruct the build system to treat subsequent warnings as errors

Signed-off-by: aoab-contributor <124131985+aoab-contributor@users.noreply.github.com>

aoab-contributor committed:
* Slightly improved parallelism in some async functions
* Resolved warnings:
   * Because this call is not awaited, execution of the current method continues before the call is completed. Consider applying the 'await' operator to the result of the call
   * This async method lacks 'await' operators and will run synchronously. Consider using the 'await' operator to await non-blocking API calls, or 'await Task.Run(...)' to do CPU-bound work on a background thread

Signed-off-by: aoab-contributor <124131985+aoab-contributor@users.noreply.github.com>

mldriscoll commented on OBB-WPF/Editor/Importer.cs:
```diff
@@ -318,26 +318,26 @@ private static void AddChapter(Chapter chapter, Chapter parentChapter, string vo
-        public Metadata metadata { get; set; }
-        public List<item> manifest { get; set; }
-        public List<itemref> spine { get; set; }
+        public Metadata Metadata { get; set; } = new Metadata();
```
The XML Serializer can be case sensitive, and the metadata/manifest/spine are lower case in the .opf files.

To get around it, you need to add tags to tell the serializer what its supposed to serialize/deserialize these properties to, e.g.

[XmlElement(ElementName = "metadata")]
[XmlArray(ElementName = "manifest")]
[XmlArray(ElementName = "spine")]

aoab-contributor replied:
Are these the only three metadata items that got messed up in this way?

I'm not familiar with the WPF stuff - as far as I can tell it relates to the non-AOAB omnibus functionality, but since I don't use it at all, I'm not sure how to test it.

Also, I'm embarrassed to admit that I don't really understand the use of code properties or how they interact with the deserializer well enough to do anything more than directly mimic your lines above... 

Does this look right? It compiles, but, again, I don't know how to test it.

```csharp

    [Serializable, XmlRoot(ElementName = "package", Namespace = "http://www.idpf.org/2007/opf")]
    [XmlType("package")]
    public class Package
    {
        [XmlElement(ElementName = "metadata")]
        public Metadata Metadata { get; set; } = new Metadata();

        [XmlArray(ElementName = "manifest")]
        public List<Item> Manifest { get; set; } = new List<Item>();

        [XmlArray(ElementName = "spine")]
        public List<Itemref> Spine { get; set; } = new List<Itemref>();
    }

```

Being optimistic, I'll go ahead and submit a new commit with these changes. But please let me know if this is wrong, or if there is more work to be done.

mldriscoll replied:
That was the only one I spotted, and yes, you've applied the tags correctly.

I've got a couple of new books to import (Tearmoon Empire and VTuber Legend have new volumes out), and I'll run those through the new code to check everything's working post-merge.

aoab-contributor replied:
Ok. Changes pushed. Please let me know if anything else needs attention.

mldriscoll commented:
Just the one actual _issue_ spotted with that one (the broken XML for the -wpf importer). 

Though as a warning, I will be turning 'treat warnings as errors' off again after merging this; Mainly because I'm well past the point where I care enough about matching Microsoft's latest code style to have it actually _enforced_ in my hobby projects.

Anyway, general question: When I put out the next releases on the JNC forums, what do you want to be credited as?

aoab-contributor commented:
Regarding warnings as errors... It's your call. If you don't want to keep it, I will just change it back on my end. In my experience, when this check is not enforced the warnings slowly creep back up over time. However I understand not wanting to bother with it for a hobby project. I will defer to your preference here.

I don't require or really want any credit. I'm mostly making these changes for my own benefit and convenience, so I'm happy enough if they can be merged at all. I actually have two more changes ready to go once this one gets merged. I think I'm more worried that you will get sick of hearing from me, than I am about receiving credit for the changes.

aoab-contributor committed:
Removed "Treat Warnings As Errors" flag from the project settings.

Signed-off-by: aoab-contributor <124131985+aoab-contributor@users.noreply.github.com>

aoab-contributor committed:
BUG FIX: WPF deserializer failure due to mismatched capitalization of data field names.

Added properties to the affected fields to adjust the behavior of the serializer.

Signed-off-by: aoab-contributor <124131985+aoab-contributor@users.noreply.github.com>

mldriscoll commented:
> I don't require or really want any credit. I'm mostly making these changes for my own benefit and convenience, so I'm happy enough if they can be merged at all. I actually have two more changes ready to go once this one gets merged. I think I'm more worried that you will get sick of hearing from me, than I am about receiving credit for the changes.

One of those upcoming changes is updating the password storage, no? Even if you don't want anything for 'mere' code cleanup, you should get credited for that. Especially as that's a function I've been forgetting to add for... well, about as long as this repo's existed...

As for things to worry about, you should be more worried I might try dumping more of my work on you than that I'll stop responding!

aoab-contributor commented:
Ok, well you can just credit me as "aoab-contributor" then. I'd like to stay as anonymous as possible.

I've submitted the password storage fixes for your review.
